hudhook/src/hooks/common.rs

```diff
@@ -116,5 +119,43 @@ pub(crate) fn imgui_wnd_proc_impl(
     let wnd_proc = imgui_renderer.wnd_proc();
     drop(imgui_renderer);
 
+    if imgui_render_loop.should_block_messages() {
+        return LRESULT(1);
+    }
+
     unsafe { CallWindowProcW(Some(wnd_proc), hwnd, umsg, WPARAM(wparam), LPARAM(lparam)) }
 }
+
+pub enum HookBackend {
+    DX11,
+    DX12
+}
+
+/// Holds information useful to the render loop which can't be retrieved from
+/// `imgui::Ui`.
+pub struct ImguiRenderLoopFlags {
+    /// Whether the hooked program's window is currently focused.
+    pub focused: bool,
+}
+
+/// Implement your `imgui` rendering logic via this trait.
+pub trait ImguiRenderLoop {
+    /// Called once at the first occurrence of the hook. Implement this to
+    /// initialize your data.
+    fn initialize(&mut self, _ctx: &mut Context) {}
+    /// Called every frame. Use the provided `ui` object to build your UI.
+    fn render(&mut self, ui: &mut Ui, flags: &ImguiRenderLoopFlags);
+
+    // If this function returns true, the WndProc function will not call the procedure of the
+    // parent window.
+    fn should_block_messages(&self) -> bool { false }
+
+    fn into_hook(self, backend: HookBackend) -> Box<dyn Hooks>
```

Comment:
I'd prefer avoiding runtime evaluations when we could statically rely on the type system. I'd instead use a trait for each backend that has a constructor, so that the user could invoke `.into_hook::<ImguiDX12Hooks>()` or something similar.